compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp

```diff
@@ -0,0 +1,158 @@
+public:
+  void runOnOperation() override;
+};
+} // namespace
+
+// Reconcile workgroup sizes across all translation infos.
+static FailureOr<SmallVector<int64_t>> reconcileWorkgroupSize(
+    ArrayRef<IREE::Codegen::TranslationInfoAttr> translationInfos) {
+  if (translationInfos.empty()) {
+    return SmallVector<int64_t>{};
+  }
+  SmallVector<int64_t> reconciledWorkgroupSize =
+      llvm::to_vector(translationInfos.front().getWorkgroupSize());
+  for (auto translationInfo : translationInfos.drop_front()) {
+    auto workGroupSize = llvm::to_vector(translationInfo.getWorkgroupSize());
+    if (workGroupSize != reconciledWorkgroupSize) {
+      return failure();
+    }
+  }
+  return reconciledWorkgroupSize;
+}
+
+// Reconcile subgroup size across all translation infos.
+static FailureOr<int64_t> reconcileSubgroupSize(
+    ArrayRef<IREE::Codegen::TranslationInfoAttr> translationInfos) {
+  if (translationInfos.empty()) {
+    return int64_t();
+  }
+  int64_t subgroupSize = translationInfos.front().getSubgroupSize();
+  for (auto translationInfo : translationInfos.drop_front()) {
+    if (subgroupSize != translationInfo.getSubgroupSize()) {
+      return failure();
+    }
+  }
+  return subgroupSize;
+}
+
+/// Helper function to retrieve the waves-per-eu value from translation info.
+static std::optional<int64_t>
+getWavesPerEu(IREE::Codegen::TranslationInfoAttr translationInfo) {
+  auto translationConfig = translationInfo.getConfiguration();
+  if (!translationConfig) {
+    return std::nullopt;
+  }
+  auto attr = translationConfig.getAs<IntegerAttr>("waves_per_eu");
+  if (!attr) {
+    return std::nullopt;
+  }
+  return attr.getValue().getSExtValue();
+}
+
+void ReconcileTranslationInfoPass::runOnOperation() {
+  auto variantOp = getOperation();
+  auto innerModuleOp = variantOp.getInnerModule();
+
+  auto exportOps = variantOp.getOps<IREE::HAL::ExecutableExportOp>();
+  if (!llvm::hasSingleElement(exportOps)) {
+    variantOp.emitOpError("reconciliation for multiple export ops unsupported");
+    return signalPassFailure();
+  }
+  auto exportOp = *exportOps.begin();
+  MLIRContext *context = &getContext();
+
+  SmallVector<IREE::Codegen::TranslationInfoAttr> translationInfos;
+  innerModuleOp->walk([&](FunctionOpInterface funcOp) {
+    auto translationInfo = getTranslationInfo(funcOp);
+    if (!translationInfo) {
+      return;
+    }
+
+    translationInfos.push_back(translationInfo);
+    // The following is moving the waves-per-eu specification from
+    // translation info into the func-like op. This is not the best
+    // place to do this, but the intent is after this pass all the
+    // lowering configs and translation infos will be deleted.
+    std::optional<int64_t> wavesPerEu = getWavesPerEu(translationInfo);
+    if (wavesPerEu) {
+      funcOp->setAttr("waves_per_eu", IntegerAttr::get(IndexType::get(context),
+                                                       wavesPerEu.value()));
```

Comment:
I'm seeing many `XXX::get` in the pass. I think we can create a `Builder builder(context)` and get the attributes from it. In this case, we can do `builder.getIndexAttr(wavesPerEu.value())` here.

Reply:
Dont want to create a builder for attributes. I want to make sure there is no IR modification in this pass (at least for now, will change in future). So Ill leave this as is.

Reply:
It is a Builder but not OpBuilder? So I think it fits what you want?

Reply:
This builder only has methods for getting attributes and types.

https://github.com/llvm/llvm-project/blob/a9d4ddd98a0bc495126027122fdca751b6841ceb/mlir/include/mlir/IR/Builders.h#L50-L205